pkg/controller/common/expectations/generations.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package expectations
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+// ExpectedStatefulSetUpdates stores StatefulSets generations that are expected in the cache,
+// following a StatefulSet update. It allows making sure we're not working with an
+// out-of-date version of the StatefulSet resource we previously updated.
+type ExpectedStatefulSetUpdates struct {
+	client      k8s.Client
+	generations map[types.NamespacedName]ResourceGeneration // per StatefulSet
+}
+
+// ResourceGeneration wraps UID and Generation for a given StatefulSet.
+type ResourceGeneration struct {
+	UID        types.UID
+	Generation int64
+}
+
+// NewExpectedStatefulSetUpdates returns an initialized ExpectedStatefulSetUpdates.
+func NewExpectedStatefulSetUpdates(client k8s.Client) *ExpectedStatefulSetUpdates {
+	return &ExpectedStatefulSetUpdates{
+		client:      client,
+		generations: make(map[types.NamespacedName]ResourceGeneration),
+	}
+}
+
+// ExpectGeneration registers the Generation of the given StatefulSets as expected.
+func (e *ExpectedStatefulSetUpdates) ExpectGeneration(statefulSet appsv1.StatefulSet) {
```

Comment:
Could we add a comment somewhere emphasising that the statefulSet we get here already has the new generation because client updates our reference? I had to spend a while thinking why we check for `>=` instead of `>` when we check the generation.